src/redisraft.c

```diff
@@ -645,6 +661,72 @@ static void handleRedisCommand(RedisRaftCtx *rr,
     raft_entry_release(entry);
 }
 
+static void handleRedisCommand(RedisRaftCtx *rr,
+                               RedisModuleCtx *ctx, RaftRedisCommandArray *cmds) {
+    /* Handle intercepted commands. We do this also on non-leader nodes or if we
+     * don't have a leader, so it's up to the commands to check these conditions
+     * if they have to.
+     */
+    if (handleInterceptedCommands(rr, ctx, cmds)) {
+        return;
+    }
+
+    /* update cmds array with the client's saved "asking" state */
+    cmds->asking = getAskingState(rr, ctx);
+
+    /* Check if MULTI/EXEC bundling is required. */
+    if (MultiHandleCommand(rr, ctx, cmds)) {
+        return;
+    }
+
+    handleRedisCommandAppend(rr, ctx, cmds);
+
+    /* reset asking state when after executing the command */
+    if (cmds->asking) {
+        setAskingState(rr, ctx, false);
+    }
+}
+
+static void redirectCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader) {
+
+    RedisModule_Assert(!raft_is_leader(rr->raft));
+
+    /* One anomaly here is that may redirect a client to the leader even for
+     * commands have no keys, which is something Redis Cluster never does. We
+     * still need to consider how this impacts clients which may not expect it.
+     */
+    replyRedirect(ctx, 0, &leader->addr);
+}
+
+static RRStatus handleNonLeaderCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, Node *leader, RaftRedisCommandArray *cmds) {
+    RedisModule_Assert(!raft_is_leader(rr->raft));
+    RedisModule_Assert(cmds);
+
+    /* reply ASK for ASKING state commands */
+    if (cmds->asking) {
+        int slot;
+        if (computeHashSlot(rr, ctx, cmds, &slot) != RR_OK) {
+            return RR_ERROR;
+        }
+        RedisModule_Assert(slot != -1); /* ASKING mode requests always have keys */
```

Comment:
I think we should not crash even if a client sends an "asking" command without a key. Client might be buggy or it might be in the development phase etc. We should reply an error in this case.